proposals/0257-string-uninitialized-initializer.md

```diff
@@ -0,0 +1,141 @@
+  ///                    initializingUTF8With: { (ptr, count) in
+  ///         ptr.initializeFrom(invalidUTF8)
+  ///         count = invalidUTF8.count
+  ///     })
+  ///     // Prints "Optional(Caf�)"
+  ///
+  ///     let s = String(unsafeUninitializedCapacity: invalidUTF8.count,
+  ///                    initializingUTF8With: { (ptr, count) in
+  ///         ptr.initializeFrom(invalidUTF8)
+  ///         count = 0
+  ///     })
+  ///     // Prints "Optional("")"
+  ///
+  /// - Parameters:
+  ///   - capacity: The number of UTF-8 code units worth of memory to allocate
+  ///       for the String.
+  ///   - initializer: A closure that initializes elements and sets the count of
+  ///       the new String
+  ///     - Parameters:
+  ///       - buffer: A buffer covering uninitialized memory with room for the
+  ///           specified number of UTF-8 code units.
+  ///       - initializedCount: Set this to the number of elements in `buffer`
+  ///           that were actually initialized by the `initializer`
+  @inlinable @inline(__always)
+  public init(
+    unsafeUninitializedCapacity capacity: Int,
+    initializingUTF8With initializer: (
+    _ buffer: UnsafeMutableBufferPointer<UInt8>,
+    _ initializedCount: inout Int
+    ) throws -> Void
+  ) rethrows
+```
+
+### Specifying a capacity
+
+The initializer takes the specific capacity that a user wants to work with as a
+parameter. The buffer passed to the closure has a count that is exactly the
+same as the specified capacity, even if the ultimate size of the new `String` is larger.
+
+### Guarantees after throwing
+
+Because `UTF8.CodeUnit` is a trivial type, there are no special considerations about the state of the buffer when an error is thrown, unlike `Array`.
+
+## Source compatibility
+
+This is an additive change to the standard library,
+so there is no effect on source compatibility.
+
+## Effect on ABI stability
+
+The new initializer will be part of the ABI, and will result in calls to a new @usableFromInline symbol being inlined into client code. Use of the new initializer is gated by @availability though, so there's no back-deployment concern.
+
+## Effect on API resilience
+
+The additional APIs will be a permanent part of the standard library,
+and will need to remain public API. 
+
+## Alternatives considered
+
+### Returning the new count from the initializer closure
+
+This is more plausible for `String` than it was for `Array`, since there's no need to deal with deinitialization of elements in partially initialized buffers (`UInt8`s are trivial). However, it was considered more valuable to match `Array`'s behavior here.
+
+### Returning a `Bool` to indicate success from the closure
+
+Requiring people to either `throw` or check in the caller for an empty `String` return if the initializing closure fails is slightly awkward, but making the initializer failable is at least as awkward, and would be inconsistent with `Array`.
+
```

Comment:
If there's discussion on the review thread, you could spend a sentence or code example about why failable is awkward. If the initializer is failable, that means there needs to be an `else` path. In practice, this basically means that every use site that doesn't want failable (which is likely the majority given that we allow for throws) has to provide some valid alternative for the else path, which is likely another implementation of the closure logic (or some generalization of it).

Reply:
Ok